File: backend/app/api/v1/routes/risks.py
# app/api/v1/routes/risks.py

from uuid import UUID

import logging

from fastapi import APIRouter, BackgroundTasks, Depends, Query, Request
from app.core.rate_limit import limiter
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_db, get_active_tenant, require_module, require_permission
from app.db.session import AsyncSessionLocal
from app.services.settings import get_ai_config
from app.schemas.risk import (
    RiskCreate, RiskUpdate, BulkImportRequest,
    AIInsightRequest, RiskStatsResponse,
)
from app.services import risk as risk_service
from app.services import ai_risk as ai_risk_service

router = APIRouter(prefix="/risks", tags=["risks"])


@router.get("")
async def list_risks(
    page: int          = Query(1, ge=1),
    page_size: int     = Query(50, ge=1, le=1000),
    risk_id: str | None   = Query(None),
    category: str | None  = Query(None),
    level: str | None     = Query(None),
    treatment: str | None = Query(None),
    owner: str | None     = Query(None),
    search: str | None    = Query(None),
    db: AsyncSession   = Depends(get_db),
    claims: dict       = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await risk_service.list_risks(
        db, tenant_id, page, page_size,
        risk_id, category, level, treatment, owner, search,
    )
    return {
        "data": result.items,
        "error": None,
        "meta": {
            "total":     result.total,
            "page":      result.page,
            "page_size": result.page_size,
            "quota":     result.quota.model_dump(),
        },
    }


@router.post("/ai")
@limiter.limit("10/minute")
async def generate_ai_insights(
    request: Request,
    payload: AIInsightRequest,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_risks")),
    _:       dict    = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await risk_service.bulk_import(db, tenant_id, payload, claims["email"])
    return {"data": result, "error": None, "meta": {}}


@router.post("/ai")
async def generate_ai_insights(
    payload: AIInsightRequest,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("generate_ai")),
    _:       dict    = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    result = await ai_risk_service.generate_insights(
        db, tenant_id, payload, claims["email"]
    )
    return {"data": result, "error": None, "meta": {}}


@router.get("/stats", response_model=None)
async def get_stats(
    category:  str | None = Query(None),
    level:     str | None = Query(None),
    treatment: str | None = Query(None),
    owner:     str | None = Query(None),
    search:    str | None = Query(None),
    db: AsyncSession      = Depends(get_db),
    claims: dict          = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    stats = await risk_service.get_stats(
        db, tenant_id, category, level, treatment, owner, search
    )
    return {"data": stats, "error": None, "meta": {}}


@router.get("/{risk_id}")
async def get_risk(
    risk_id: str,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    risk = await risk_service.get_risk(db, tenant_id, risk_id)
    return {"data": risk, "error": None, "meta": {}}


logger = logging.getLogger(__name__)


async def _auto_run_ai(tenant_id: UUID, risk_id: str, user_email: str) -> None:
    async with AsyncSessionLocal() as db:
        try:
            ai_cfg = await get_ai_config(db, tenant_id)
            if not ai_cfg['enabled'] or not ai_cfg['auto_run']:
                return
            req = AIInsightRequest(
                target='selected',
                risk_ids=[risk_id],
                confidence=ai_cfg['confidence'],
                overwrite=False,
            )
            await ai_risk_service.generate_insights(db, tenant_id, req, user_email)
            await db.commit()
        except Exception as exc:
            logger.error('auto_run_ai failed | risk=%s | %s', risk_id, exc)


@router.post("")
@limiter.limit("60/minute")
async def create_risk(
    request:          Request,
    payload:          RiskCreate,
    background_tasks: BackgroundTasks,
    db:               AsyncSession = Depends(get_db),
    claims:           dict         = Depends(require_permission("manage_risks")),
    _:                dict         = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    risk = await risk_service.create_risk(db, tenant_id, payload, claims["email"])
    background_tasks.add_task(_auto_run_ai, tenant_id, risk.id, claims["email"])
    return {"data": risk, "error": None, "meta": {}}


@router.patch("/{risk_id}")
@limiter.limit("60/minute")
async def update_risk(
    request: Request,
    risk_id: str,
    payload: RiskUpdate,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_risks")),
    _:       dict    = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    risk = await risk_service.update_risk(db, tenant_id, risk_id, payload, claims["email"])
    return {"data": risk, "error": None, "meta": {}}


@router.delete("/{risk_id}")
@limiter.limit("60/minute")
async def delete_risk(
    request: Request,
    risk_id: str,
    db: AsyncSession = Depends(get_db),
    claims: dict     = Depends(require_permission("manage_risks")),
    _:       dict    = Depends(require_module("risk")),
):
    tenant_id = UUID(claims["active_tenant_id"])
    await risk_service.delete_risk(db, tenant_id, risk_id, claims["email"])
    return {
        "data": {"message": f"Risk {risk_id} moved to recycle bin"},
        "error": None,
        "meta": {},
    }

File: frontend/src/services/risks.ts
// src/services/risks.ts

import api from './api';
import { apiGet, apiPost, apiPatch, apiDelete } from './api';
import type { ApiResponse } from '../types/api';
import type {
  Risk, RiskCreate, RiskUpdate, RiskListMeta,
  BulkImportRow, BulkImportResult,
  AIInsightRequest, AIInsightResult,
  RecycleBinItem, RiskStats,
} from '../types/risk';

export interface ListRisksParams {
  page?:      number;
  page_size?: number;
  risk_id?:   string;
  category?:  string;
  level?:     string;
  treatment?: string;
  owner?:     string;
  search?:    string;
  undecided?: boolean;
}

export interface RiskListResponse {
  items: Risk[];
  meta:  RiskListMeta;
}

export async function listRisks(params: ListRisksParams = {}): Promise<RiskListResponse> {
  const query = new URLSearchParams();
  if (params.page)      query.set('page',      String(params.page));
  if (params.page_size) query.set('page_size', String(params.page_size));
  if (params.risk_id)   query.set('risk_id',   params.risk_id);
  if (params.category)  query.set('category',  params.category);
  if (params.level)     query.set('level',      params.level);
  if (params.treatment) query.set('treatment',  params.treatment);
  if (params.owner)     query.set('owner',      params.owner);
  if (params.search)    query.set('search',     params.search);
  if (params.undecided) query.set('undecided',  'true');

  const qs  = query.toString();
  const url = `/api/v1/risks${qs ? `?${qs}` : ''}`;
  const envelope = await api.get<ApiResponse<Risk[]>>(url);
  if (envelope.data.error) throw new Error(envelope.data.error);

  return {
    items: envelope.data.data ?? [],
    meta:  envelope.data.meta as unknown as RiskListMeta,
  };
}

export async function getRisk(id: string): Promise<Risk> {
  return apiGet<Risk>(`/api/v1/risks/${id}`);
}

export async function createRisk(payload: RiskCreate): Promise<Risk> {
  return apiPost<Risk>('/api/v1/risks', payload);
}

export async function updateRisk(id: string, payload: RiskUpdate): Promise<Risk> {
  return apiPatch<Risk>(`/api/v1/risks/${id}`, payload);
}

export async function deleteRisk(id: string): Promise<void> {
  await apiDelete(`/api/v1/risks/${id}`);
}

export async function bulkImport(rows: BulkImportRow[]): Promise<BulkImportResult> {
  return apiPost<BulkImportResult>('/api/v1/risks/import', { rows });
}

export async function generateAI(payload: AIInsightRequest): Promise<AIInsightResult> {
  return apiPost<AIInsightResult>('/api/v1/risks/ai', payload);
}

export interface StatsParams {
  category?:  string;
  level?:     string;
  treatment?: string;
  owner?:     string;
  search?:    string;
  undecided?: boolean;
}

export async function getStats(params: StatsParams = {}): Promise<RiskStats> {
  const qs = new URLSearchParams();
  if (params.category)  qs.set('category',  params.category);
  if (params.level)     qs.set('level',      params.level);
  if (params.treatment) qs.set('treatment',  params.treatment);
  if (params.owner)     qs.set('owner',      params.owner);
  if (params.search)    qs.set('search',     params.search);
  const query = qs.toString();
  return apiGet<RiskStats>(`/api/v1/risks/stats${query ? `?${query}` : ''}`);
}

export async function listBin(itemType?: string): Promise<RecycleBinItem[]> {
  const qs = itemType ? `?item_type=${itemType}` : '';
  return apiGet<RecycleBinItem[]>(`/api/v1/recycle${qs}`);
}

export async function restoreItem(binId: string): Promise<void> {
  await apiPost(`/api/v1/recycle/${binId}/restore`, {});
}

export async function permanentDelete(binId: string): Promise<void> {
  await apiDelete(`/api/v1/recycle/${binId}`);
}